File: app/src/main/java/com/example/dell/chargehelper/notifications/ApplicationNotificator.java
package com.example.dell.chargehelper.notifications;

import android.app.Activity;
import android.app.AlarmManager;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.media.AudioAttributes;
import android.net.Uri;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.RequiresApi;
import com.example.dell.chargehelper.R;
import com.example.dell.chargehelper.SettingsProvider;
import com.example.dell.chargehelper.helpers.TimeHelper;
import com.example.dell.chargehelper.UserMessage;

import java.util.Date;

public class ApplicationNotificator implements INotificator
{
    private final Activity activity;
    private final Uri SoundUri;

    ApplicationNotificator(Activity activity) {
        this.activity = activity;
        SoundUri = Uri.parse("android.resource://" + activity.getPackageName() + "/" + R.raw.carhorn4);
    }

    @Override
    public void scheduleCarChargedNotification(long millisToEvent) {
        long millisToNotify = getMillisToNotify(millisToEvent);
        String description = getNotificationDescription(millisToEvent);

        Notification notification = getCarChargedNotification(
            activity.getString(R.string.car_charged_title),
            description
        );

        PendingIntent pendingIntent = getNotificationIntent(notification);
        scheduleNotification(millisToNotify, pendingIntent);
        UserMessage.showSnackbar(activity, description);
    }

    private long getMillisToNotify(long millisToEvent) {
        SettingsProvider settingsProvider = new SettingsProvider(activity);
        long millisToNotify = TimeHelper.convertMinutesToMs(settingsProvider.getApplicationReminderMinutes());
        if (millisToEvent > millisToNotify)
            return millisToEvent - millisToNotify;
        return millisToEvent;
    }

    private String getNotificationDescription(long duration) {
        Date chargedAtTime = TimeHelper.toDate(TimeHelper.addToNow(duration));
        String descriptionTemplate = activity.getString(R.string.car_charged_time_desc);
        return String.format(descriptionTemplate, TimeHelper.formatAsHoursWithMinutes(chargedAtTime));
    }

    private static final String CAR_CHARGED_NOTIFICATION_CHANNEL_ID = "46578";
    private Notification getCarChargedNotification(String title, String description) {
        Notification.Builder builder = new Notification.Builder(activity)
                .setContentTitle(title)
                .setContentText(description)
                .setSmallIcon(R.drawable.ic_timer_24px)
                .setSound(SoundUri);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O){
            NotificationChannel channel = getNotificationChannel();
            registerNotificationChannel(channel);

            builder.setChannelId(CAR_CHARGED_NOTIFICATION_CHANNEL_ID);
        }

        return builder.build();
    }


    @RequiresApi(api = Build.VERSION_CODES.O)
    @NonNull
    private NotificationChannel getNotificationChannel() {
        NotificationChannel channel = new NotificationChannel(CAR_CHARGED_NOTIFICATION_CHANNEL_ID,
                activity.getString(R.string.car_charged_title),
                NotificationManager.IMPORTANCE_DEFAULT);

        channel.setDescription(activity.getString(R.string.car_charged_channel_descr));
        channel.setLockscreenVisibility(Notification.VISIBILITY_PUBLIC);

        AudioAttributes attributes = new AudioAttributes.Builder()
                .setUsage(AudioAttributes.USAGE_NOTIFICATION)
                .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
                .build();

        channel.setSound(SoundUri, attributes);
        return channel;
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    private void registerNotificationChannel(NotificationChannel channel) {
        NotificationManager notificationManager = (NotificationManager)activity.getSystemService(Context.NOTIFICATION_SERVICE);
        assert notificationManager != null;

        notificationManager.createNotificationChannel(channel);
    }

    private PendingIntent getNotificationIntent(Notification notification) {
        Intent notificationIntent = new Intent(activity, CarChargedAlarmReceiver.class);
        notificationIntent.putExtra(CarChargedAlarmReceiver.NOTIFICATION_ID, 1);
        notificationIntent.putExtra(CarChargedAlarmReceiver.NOTIFICATION, notification);
        return PendingIntent.getBroadcast(activity, 0, notificationIntent, PendingIntent.FLAG_UPDATE_CURRENT);
    }

    private void scheduleNotification(long millis, PendingIntent pendingIntent) {
        AlarmManager alarmManager = (AlarmManager)activity.getSystemService(Context.ALARM_SERVICE);
        long triggerAt = TimeHelper.addToNow(millis);

        assert alarmManager != null;
        alarmManager.set(AlarmManager.RTC_WAKEUP, triggerAt, pendingIntent);
    }
}

File: EvChargeTimer/app/src/main/java/com/chebuso/chargetimer/notifications/application/ApplicationNotificator.kt
package com.chebuso.chargetimer.notifications.application

import android.app.Activity
import com.chebuso.chargetimer.R
import com.chebuso.chargetimer.shared.UserMessage.showSnackbar
import com.chebuso.chargetimer.shared.helpers.TimeHelper
import com.chebuso.chargetimer.shared.helpers.TimeHelper.formatAsShortTime
import com.chebuso.chargetimer.shared.helpers.TimeHelper.toDate
import com.chebuso.chargetimer.notifications.INotificator
import com.chebuso.chargetimer.settings.ISettingsReader
import java.util.Date


class ApplicationNotificator internal constructor(
    private val settingsProvider: ISettingsReader,
    private val registrar: NotificationChannelRegistrar,
    private val alarmScheduler: NotificationAlarmScheduler,
    private val activity: Activity
) : INotificator {

    override fun scheduleCarChargedNotification(millisToEvent: Long) {
        val triggerAtMillis = getTriggerAtMillis(millisToEvent)
        val chargedAtMillis = getChargedAtMillis(millisToEvent)

        val channelId = registrar.registerCarChargedNotificationChannel()
        alarmScheduler.schedule(AlarmItem(
            triggerAtMillis,
            chargedAtMillis,
            CAR_CHARGED_NOTIFICATION_ID,
            channelId,
        ))

        val text = getSnackbarText(toDate(chargedAtMillis), toDate(triggerAtMillis))
        showSnackbar(activity, text)
    }

    private fun getTriggerAtMillis(millisToEvent: Long): Long{
        val millisToNotify = getMillisToNotify(millisToEvent)
        return TimeHelper.now() + millisToNotify
    }

    private fun getChargedAtMillis(millisToEvent: Long): Long = TimeHelper.now() + millisToEvent

    private fun getMillisToNotify(millisToEvent: Long): Long {
        val minutesToRemind = settingsProvider.getApplicationReminderMinutes()
        val millisToNotify = TimeHelper.convertMinutesToMs(minutesToRemind)
        return if (millisToEvent > millisToNotify) millisToEvent - millisToNotify else millisToEvent
    }

    private fun getSnackbarText(chargedAt: Date, triggerAt: Date): String {
        val descriptionTemplate = activity.getString(R.string.car_charged_time_snackbar_desc)
        return String.format(descriptionTemplate,
            chargedAt.formatAsShortTime(),
            triggerAt.formatAsShortTime())
    }

    companion object {
        private val TAG = ApplicationNotificator::class.java.simpleName
        private const val CAR_CHARGED_NOTIFICATION_ID = 204
    }
}

